cupy/linalg/_eigenvalue.py

```diff
@@ -141,7 +199,53 @@ def eigh(a, UPLO='L'):
         return _syevd(a, UPLO, True)
 
 
-# TODO(okuta): Implement eigvals
+def eig(a):
+    """
+    Return the eigenvalues and eigenvectors of a matrix.
+
+    Returns two objects, a 1-D array containing the eigenvalues of `a`, and
+    a 2-D square array or matrix (depending on the input type) of the
+    corresponding eigenvectors (in columns).
+
+    Args:
+        a (cupy.ndarray): A symmetric 2-D square matrix ``(M, M)`` or a batch
+            of symmetric 2-D square matrices ``(..., M, M)``.
+    Returns:
+        tuple of :class:`~cupy.ndarray`:
+            Returns a tuple ``(w, v)``. ``w`` contains eigenvalues and
+            ``v`` contains eigenvectors. ``v[:, i]`` is an eigenvector
+            corresponding to an eigenvalue ``w[i]``. For batch input,
+            ``v[k, :, i]`` is an eigenvector corresponding to an eigenvalue
+            ``w[k, i]`` of ``a[k]``.
+    Notes:
+        There is no guarantee of the order of the eigenvalues:
+        it can even be different from ``numpy.linalg.eig``.
+
+    .. warning::
+        This function calls one or more cuSOLVER routine(s) which may yield
+        invalid results if input conditions are not met.
+        To detect these invalid results, you can set the `linalg`
+        configuration to a value that is not `ignore` in
+        :func:`cupyx.errstate` or :func:`cupyx.seterr`.
+
+    .. seealso:: :func:`numpy.linalg.eig`
+    """
+    _util._assert_stacked_2d(a)
+    _util._assert_stacked_square(a)
+
+    if a.size == 0:
+        _, v_dtype = _util.linalg_common_type(a)
+        w = cupy.empty(a.shape[:-1], v_dtype)
+        v = cupy.empty(a.shape, v_dtype)
+        return w, v
+
+    if a.ndim == 2:
+        return _geev(a, True)
+
+    work = [_geev(a[ind, :, :], True) for ind in numpy.ndindex(a.shape[:-2])]
+    w = cupy.stack([x[0] for x in work])
+    v = cupy.stack([x[1] for x in work])
```

Comment:
This uses a Python loop to call geek. I suggest:
1. either we create an issue to track this overhead before merging, or
2. we just move this loop to C++ in this PR (see [cusolver.pyx](https://github.com/cupy/cupy/blob/main/cupyx/cusolver.pyx) for batched SVD/QR examples)